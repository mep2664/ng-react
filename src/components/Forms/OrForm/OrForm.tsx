import * as React from "react";
import { OrButton } from "../../";
import { bgColor, borderColor, emphasisType, fontColor } from "../../../theme";
import styled from "styled-components";

interface IEmphasis {
    bgEmphasis: emphasisType;
    fontEmphasis: emphasisType;
}

const Wrapper = styled.div<IEmphasis>`
    display: grid;
    grid-template-areas: 'main' 'footer';
    grid-gap: 15px;
    grid-template-rows: auto 60px;
    min-height: 350px;
    width: 400px;
    padding: 25px;
    color: ${({ fontEmphasis }) => fontColor[fontEmphasis]};
    background-color: ${({ bgEmphasis }) => bgColor[bgEmphasis]};
    border: 1px solid ${({ bgEmphasis }) => borderColor[bgEmphasis]};
    border-radius: 5px;
    position: relative;
`;

const FormContainer = styled.div`
    grid-area: main;
`;

const ButtonContainer = styled.div`
    grid-area: footer;
    display: flex;
    align-items: flex-end;
`;

const SubmitOverlay = styled.div`
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    background-color: ${bgColor["Overlay"]};
    cursor: wait;
`;

export interface IOrFormProps {
    leftFormID: string;
    rightFormID: string;
    leftButtonCaption: string;
    rightButtonCaption: string;
    bgEmphasis?: emphasisType;
    fontEmphasis?: emphasisType;
    leftForm: (props?: any) => JSX.Element;
    rightForm: (props?: any) => JSX.Element;
};

export const OrForm: React.FC<IOrFormProps> = ({ leftFormID, rightFormID, leftButtonCaption, rightButtonCaption, leftForm, rightForm, bgEmphasis = "Primary", fontEmphasis = "Secondary" }) => {
    const [isLeftActive, setLeftActive] = React.useState<boolean>(true);
    const [isSubmitting, setSubmitting] = React.useState<boolean>(false);

    const handleLeftOnClick = (e: React.MouseEvent<HTMLButtonElement>) => {
        if (!isLeftActive) {
            e.preventDefault();
        }
        setLeftActive(true);
    }

    const handleRightOnClick = (e: React.MouseEvent<HTMLButtonElement>) => {
        if (isLeftActive) {
            e.preventDefault();
        }
        setLeftActive(false);
    }

    return (
        <Wrapper bgEmphasis={bgEmphasis} fontEmphasis={fontEmphasis}>
            <FormContainer>
                {isLeftActive ? leftForm({ onSubmitChange: (isSubmitting: boolean) => setSubmitting(isSubmitting), emphasis: bgEmphasis })
                    : rightForm({ onSubmitChange: (isSubmitting: boolean) => setSubmitting(isSubmitting), emphasis: bgEmphasis })}
            </FormContainer>
            <ButtonContainer>
                <OrButton
                    leftCaption={leftButtonCaption}
                    rightCaption={rightButtonCaption}
                    leftType={isLeftActive ? "submit" : "button"}
                    rightType={isLeftActive ? "button" : "submit"}
                    leftForm={leftFormID}
                    rightForm={rightFormID}
                    leftOnClick={handleLeftOnClick}
                    rightOnClick={handleRightOnClick}
                />
            </ButtonContainer>
            {isSubmitting && <SubmitOverlay role="alert" aria-busy="true" aria-live="polite" />}
        </Wrapper>
    );
}



