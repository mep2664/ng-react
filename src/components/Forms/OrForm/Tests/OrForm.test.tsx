import React from "react"
import { fireEvent, render, wait } from "@testing-library/react"
import "@testing-library/jest-dom/extend-expect";
import { OrForm } from "../OrForm"

const leftFormTitle = "LF_TITLE";
const rightFormTitle = "RF_TITLE";
const props = {
    leftFormID: "LEFT_FORM",
    rightFormID: "RIGHT_FORM",
    leftButtonCaption: "LEFT_BTN",
    rightButtonCaption: "RIGHT_BTN",
    leftForm: (props: any) => <form title={leftFormTitle} onSubmit={() => props.onSubmitChange(true)} />,
    rightForm: (props: any) => <form title={rightFormTitle} onSubmit={() => props.onSubmitChange(true)} />,
};

describe("Tests for the OrForm component", () => {
    beforeEach(() => {
        jest.resetAllMocks();
    });

    test("Component matches snapshot", () => {
        const { container } = render(
            <OrForm {...props} />
        );
        expect(container).toMatchSnapshot();
    });

    test("Left form is displayed on render", () => {
        const { getByTitle } = render(
            <OrForm {...props} />
        );
        expect(getByTitle(leftFormTitle)).toBeDefined();
    });

    test("Right form is displayed after right button click", () => {
        const { getByText, getByTitle, queryByTitle } = render(
            <OrForm {...props} />
        );
        expect(queryByTitle(rightFormTitle)).toBeNull();
        fireEvent.click(getByText(props.rightButtonCaption));
        expect(getByTitle(rightFormTitle)).toBeDefined();
    });

    test("Left form is displayed after left button click", () => {
        const { getByText, getByTitle, queryByTitle } = render(
            <OrForm {...props} />
        );
        fireEvent.click(getByText(props.rightButtonCaption));
        expect(queryByTitle(leftFormTitle)).toBeNull();
        fireEvent.click(getByText(props.leftButtonCaption));
        expect(getByTitle(leftFormTitle)).toBeDefined();
    });

    test("Right button type is correct after form changes", () => {
        const { getByText } = render(
            <OrForm {...props} />
        );
        const rightBtn = getByText(props.rightButtonCaption);
        expect(rightBtn).toHaveAttribute("type", "button");
        fireEvent.click(rightBtn);
        expect(rightBtn).toHaveAttribute("type", "submit");
        fireEvent.click(getByText(props.leftButtonCaption));
        expect(rightBtn).toHaveAttribute("type", "button");
    });

    test("Left button type is correct after form changes", () => {
        const { getByText } = render(
            <OrForm {...props} />
        );
        const leftBtn = getByText(props.leftButtonCaption);
        const rightBtn = getByText(props.rightButtonCaption);
        expect(leftBtn).toHaveAttribute("type", "submit");
        fireEvent.click(rightBtn);
        expect(leftBtn).toHaveAttribute("type", "button");
        fireEvent.click(leftBtn);
        expect(leftBtn).toHaveAttribute("type", "submit");
        fireEvent.click(rightBtn);
        expect(leftBtn).toHaveAttribute("type", "button");
    });

    test("Loader is displayed after left form submit", async () => {
        const { getByTitle, getByRole } = render(
            <OrForm {...props} />
        );
        await wait();
        fireEvent.submit(getByTitle(leftFormTitle));
        const loading = getByRole("alert");
        expect(loading).toBeDefined();
        expect(loading).toHaveAttribute("aria-busy", "true");
    });

    test("Loader is displayed after right form submit", async () => {
        const { getByText, getByTitle, getByRole } = render(
            <OrForm {...props} />
        );
        await wait();
        fireEvent.click(getByText(props.rightButtonCaption));
        fireEvent.submit(getByTitle(rightFormTitle));
        const loading = getByRole("alert");
        expect(loading).toBeDefined();
        expect(loading).toHaveAttribute("aria-busy", "true");
    });
});
