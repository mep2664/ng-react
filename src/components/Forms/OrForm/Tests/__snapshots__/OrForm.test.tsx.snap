// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for the OrForm component Component matches snapshot 1`] = `
<div>
  <div
    class="sc-hzDkRC fjJNAb"
  >
    <div
      class="sc-jhAzac dLhCdB"
    >
      <form
        title="LF_TITLE"
      />
    </div>
    <div
      class="sc-fBuWsC dAloEz"
    >
      <div
        class="sc-bwzfXH jElIFP"
      >
        <button
          class="sc-bxivhb kZMPJz"
          form="LEFT_FORM"
          type="submit"
        >
          LEFT_BTN
        </button>
        <span
          class="sc-htpNat fEbLPN"
        >
          or
        </span>
        <button
          class="sc-ifAKCX ewbGBU"
          form="RIGHT_FORM"
          type="button"
        >
          RIGHT_BTN
        </button>
      </div>
    </div>
  </div>
</div>
`;
